### PR TITLE
[PyPI] Remove colorama from hard dependencies

### PR DESCRIPTION
**Pull request by archibate**
Related issue = #1457

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----
I'm afraid `colorama` will also cause issues like #1457... anyway, less dep, less possible to error :)

### CODE REVIEW

**Comment by k-ye** [comment]
I saw that #1477 removed `pytest`, which, although breaks `ti test` for non-developers, but probably makes sense anyway because users don't need to care about tests. But I do have a general question for this: Are we going to provide a mock for any dependency package just because the users couldn't install it? (Personally I don't prefer this because it just keeps adding  maintenance costs.)

**Comment by archibate** [comment]
> mock for any dependency package just because the users couldn't install it?

We will for simple deps like colorama, but we won't mock for fundemental deps like numpy or pybind11.

> maintenance costs

-1, wasn't it just a dummy color.py? maintain? I think we will never touch it any more.

File: setup.py [hunk 1]
```diff
@@ -45,5 +44,6 @@
                  },
                  classifiers=classifiers,
                  has_ext_modules=lambda: True)
+>>>>>>> master
```
Comment (rexwangcc) [inline setup.py:47]: seems like a leftover from rebasing?

File: python/taichi/lang/util.py [hunk 2]
```diff
@@ -1,14 +1,13 @@
 from .core import taichi_lang_core
 import numpy as np
-import os
 
 _has_pytorch = False
-
-_env_torch = os.environ.get('TI_ENABLE_TORCH', '1')
+_env_torch = __import__('os').environ.get('TI_ENABLE_TORCH', '1')
```
Comment (rexwangcc) [inline python/taichi/lang/util.py:5]: just curious: why this change here?

Comment (archibate) [reply python/taichi/lang/util.py:5]: Since we use the `from .util import *` pattern... we will get `ti.os` at the end.

Comment (rexwangcc) [reply python/taichi/lang/util.py:5]: I see, so this is to resolve the potential monkey patching problem caused by `import *`? That makes sense to me. Again, I really hope we could get rid of `import *` patterns wherever possible : /

**Comment by rexwangcc** [comment]
> I saw that #1477 removed `pytest`, which, although breaks `ti test` for non-developers, probably makes sense anyway because users don't need to care about tests. But I do have a general question for this: Are we going to provide a mock for any dependency package just because the users couldn't install it? (Personally I don't prefer this because it just keeps adding maintenance costs.)

That's actually a good point. A potential compromise might be move all of the deps that are trivial to end-users to a `extra_deps` field so they could still be installed by `pip install taichi[trivial_crap]` while most of end-users won't need to care about it. And +1 to you point about `adding maintenance costs`, since each time the interfaces of these deps change, we will have to update the mocks 🔨

**Comment by archibate** [comment]
> pip install taichi[trivial_crap]

Sounds cool! How to implement this in `setup.py`?

**Comment by Rullec** [comment]
Nice work! Please let me read it carefully tomorrow morning. Good night!

**Comment by rexwangcc** [comment]
> > pip install taichi[trivial_crap]
> 
> Sounds cool! How to implement this in `setup.py`?

I think we want to follow something like https://github.com/zalando/connexion/blob/bed4b95205205861bd6014282d0f61d50231d3ac/setup.py#L98, which makes something like `pip install taichi[test-suite]` feasible. 

There are docs explaining it here: https://setuptools.readthedocs.io/en/latest/setuptools.html#declaring-extras-optional-features-with-their-own-dependencies

We could do it later in a separate PR since there are problems with this method too. (e.g. I can see `pip install xx[yy]` breaks Zsh shells already, might need something like `pip install "xx[yy]"`)

File: python/taichi/lang/util.py (hunk 2 above)
Comment (Rullec) [reply python/taichi/lang/util.py:5]: > I see, so this is to resolve the potential monkey patching problem caused by `import *`? That makes sense to me. Again, I really hope we could get rid of `import *` patterns wherever possible : /

Agree, ``import *`` can be kinds of dangerous potentially

**Comment by Rullec** [comment]
LGTM!

**Comment by yuanming-hu** [comment]
Thanks, but TBH I don't really understand why we need to get rid of `colorama`. It's not a heavy dependency and there's no clear evidence that it will bring us issues. I kind of agree with @k-ye that maybe we should simply depend on `colorama` for simplicity. I don't understand #1498 as well.  I believe simplicity matters more.

> > maintenance costs
> 
> -1, wasn't it just a dummy color.py? maintain? I think we will never touch it any more.

First, please be courteous. Second, every line of code has its maintenance cost. I don't think "we will never touch it anymore".
